File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

codes :: [[String] -> [String]]
codes = [wink, doubleBlink, closeEyes, jump, reverse]
  where wink = (++ ["wink"])
        doubleBlink = (++ ["double blink"])
        closeEyes = (++ ["close your eyes"])
        jump = (++ ["jump"])

handshake :: Int -> [String]
handshake = foldl apply [] . zip codes . bits
  where apply lst (act, True) = act lst
        apply lst _ = lst
        bits 0 = []
        bits n = odd n : bits (n `div` 2)

File: clojure/secret-handshake/src/secret_handshake.clj
(ns secret-handshake)

(def codes {0 #(conj % "wink")
            1 #(conj % "double blink")
            2 #(conj % "close your eyes")
            3 #(conj % "jump")
            4 reverse})

(defn commands [num]
  (->> (range 5)
       (filter #(bit-test num %))
       (reduce #((get codes %2) %1) [])))
